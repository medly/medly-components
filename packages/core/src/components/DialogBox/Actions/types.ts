export type DialogBoxActionUserProps = {
    /** Use this to align actions horizontally */
    alignItems?: 'left' | 'center' | 'right';
};
