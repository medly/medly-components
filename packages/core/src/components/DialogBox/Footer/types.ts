export type DialogBoxFooterProps = {
    /** Use this to align actions horizontally */
    alignItems?: 'left' | 'center' | 'right';
};
