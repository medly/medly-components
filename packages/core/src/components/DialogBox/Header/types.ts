import { HTMLProps, WithThemeProp } from '@medly-components/utils';
export interface DialogBoxHeaderProps extends HTMLProps<HTMLDivElement>, WithThemeProp {
    id: string;
}
