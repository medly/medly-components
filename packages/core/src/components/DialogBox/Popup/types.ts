import { HTMLProps } from '@medly-components/utils';

export interface DialogBoxPopupProps extends HTMLProps<HTMLDivElement> {
    minWidth?: string;
    minHeight?: string;
    open?: boolean;
}
