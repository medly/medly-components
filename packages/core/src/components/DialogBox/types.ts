import { HTMLProps, WithStyle } from '@medly-components/utils';
import { DialogBoxActionUserProps } from './Actions/types';
import { DialogBoxPopupProps } from './Popup/types';

export interface DialogBoxProps extends HTMLProps<HTMLDivElement> {
    /** Shows modal only when this prop is true */
    open?: boolean;
    /** Function to be called on closing modal */
    onCloseModal?: () => void;
    /** Min width in px/rem/% (1rem = 10px) */
    minWidth?: string;
    /** Min height in px/rem/% (1rem = 10px) */
    minHeight?: string;
    /** Set it true to close the modal on click outside of popup */
    shouldCloseOnOutsideClick?: boolean;
}

export interface DialogBoxBackgroundProps {
    open?: boolean;
}

export interface DialogBoxStaticProps {
    Popup: React.FC<DialogBoxPopupProps> & WithStyle;
    Header: React.FC & WithStyle;
    Content: React.FC & WithStyle;
    Actions: React.FC<DialogBoxActionUserProps> & WithStyle;
}
export interface DialogBoxContextType {
    id: string;
}
