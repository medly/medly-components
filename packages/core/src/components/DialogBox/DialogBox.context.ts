import { createContext } from 'react';
import { DialogBoxContextType } from './types';

// default values are for testing purposes
export const defaultValues: DialogBoxContextType = {
    id: 'default-id'
};

export const DialogBoxContext = createContext<DialogBoxContextType>(defaultValues);
