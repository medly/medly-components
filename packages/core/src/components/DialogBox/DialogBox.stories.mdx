import { DialogBox } from './DialogBox';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import * as stories from './DialogBox.stories';

<Meta title="Core" component={DialogBox} parameters={{ jest: ['DialogBox.test.tsx'] }} />

# DialogBox

`DialogBox` are windows that appear in front of app content to give critical information, request a decision, or involve multiple tasks.  It consists of a Header, Content, and Action.

You can play with the component in the canvas tab.

```typescript
<DialogBox open={openState} onCloseDialogBox={changeDialogBoxState(false)}>
    <DialogBox.Header>Add User</DialogBox.Header>
    <DialogBox.Content>
        <Input id="name-input" type="text" fullWidth label="Name" placeholder="Enter your Name" />
    </DialogBox.Content>
    <DialogBox.Actions>
        <Button variant="outlined">Cancel</Button>
    </DialogBox.Actions>
</DialogBox>
```

Shadow appears on the Header and Actions section if you scroll through the content.

To close the dialog box, you can either:

1. Click the cross icon.
2. Press `Esc` key.
3. If the `shouldCloseOnOutsideClick` prop is provided, then the dialog box will close on clicking outside the dialog box.

<Preview withToolbar>
    <Story name="DialogBox">{stories.Basic()}</Story>
</Preview>

## Props

This component offers the following props mentioned in the table below:

<Props of={DialogBox} />

## DialogBox Action Props

The DialogBox action prop is given in the table below:

<Props of={stories.DialogBoxActionProps} />

## DialogBoxTheme

This component offers the following themes mentioned in the table below:

<Props of={stories.ThemeInterface} />
