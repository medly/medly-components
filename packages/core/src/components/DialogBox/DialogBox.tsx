import { useCombinedRefs, useKeyPress, WithStyle } from '@medly-components/utils';
import { memo, forwardRef, useRef, useCallback, useEffect, useState } from 'react';
import Actions from './Actions';
import Content from './Content';
import { DialogBoxContext } from './DialogBox.context';
import { DialogBoxBackgroundStyled } from './DialogBox.styled';
import Header from './Header';
import Popup from './Popup';
import { DialogBoxProps, DialogBoxStaticProps } from './types';
import type { FC } from 'react';
import CloseIcon from '../Modal/CloseIcon';

const Component: FC<DialogBoxProps> = memo(
    forwardRef((props, ref) => {
        const {
                id,
                open,
                onCloseModal,
                children,
                minWidth,
                shouldCloseOnOutsideClick,
                minHeight,
                showCloseIcon = false,
                ...restProps
            } = props,
            isEscPressed = useKeyPress('Escape'),
            dialogBoxRef = useCombinedRefs<HTMLDivElement>(ref, useRef(null)),
            [shouldRender, setShouldRender] = useState(open);

        const handleBackgroundClick = useCallback(() => {
            shouldCloseOnOutsideClick && onCloseModal && onCloseModal();
        }, [shouldCloseOnOutsideClick, onCloseModal]);

        useEffect(() => {
            if (open) {
                setShouldRender(true);
                document.body.style.overflow = 'hidden';
            }
            return () => {
                document.body.style.overflow = 'unset';
            };
        }, [open]);

        const handleAnimationEnd = useCallback(() => {
            if (!open) {
                setShouldRender(false);
                document.body.style.overflow = 'unset';
            }
        }, [open]);

        useEffect(() => {
            open && isEscPressed && onCloseModal && onCloseModal();
        }, [open, isEscPressed]);

        const handleCloseModal = useCallback(() => {
            onCloseModal && onCloseModal();
        }, [onCloseModal]);

        return shouldRender ? (
            <DialogBoxBackgroundStyled {...{ ...restProps, id, open }} onClick={handleBackgroundClick}>
                <Popup ref={dialogBoxRef} id={`${id}-popup`} onAnimationEnd={handleAnimationEnd} {...{ minWidth, minHeight, open }}>
                    {showCloseIcon && (
                        <CloseIcon
                            id={`${id}-close-button`}
                            title={`${id}-close-icon`}
                            size="M"
                            variant="solid"
                            data-testid={`${id}-close-button`}
                            onClick={handleCloseModal}
                        />
                    )}
                    <DialogBoxContext.Provider value={{ id: id! }}>{children}</DialogBoxContext.Provider>
                </Popup>
            </DialogBoxBackgroundStyled>
        ) : null;
    })
);

Component.defaultProps = {
    id: 'medly-dialog-box',
    open: false,
    shouldCloseOnOutsideClick: false,
    showCloseIcon: false
};
Component.displayName = 'DialogBox';
export const DialogBox: FC<DialogBoxProps> & WithStyle & DialogBoxStaticProps = Object.assign(Component, {
    Style: DialogBoxBackgroundStyled,
    Header,
    Popup,
    Content,
    Actions
});
