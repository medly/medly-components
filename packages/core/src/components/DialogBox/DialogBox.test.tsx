import { fireEvent, render } from '@test-utils';
import React from 'react';
import { DialogBox } from './DialogBox';
import { DialogBoxProps } from './types';

const dialogBoxRenderer = ({
    open = false,
    onCloseModal = jest.fn(),
    minWidth,
    minHeight,
    shouldCloseOnOutsideClick = false
}: DialogBoxProps) =>
    render(
        <DialogBox {...{ open, onCloseModal, minHeight, minWidth, shouldCloseOnOutsideClick }}>
            <DialogBox.Header>
                <p>Demo Header</p>
            </DialogBox.Header>
            <DialogBox.Content>
                Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry standard dummy
                text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It
                has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was
                popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop
                publishing software like Aldus PageMaker including versions of Lorem Ipsum
            </DialogBox.Content>
            <DialogBox.Actions>Demo Actions</DialogBox.Actions>
        </DialogBox>
    );
describe('DialogBox component', () => {
    it('should render properly when it is open', () => {
        const { container } = dialogBoxRenderer({ open: true, minWidth: '200px', minHeight: '200px' });
        expect(container).toMatchSnapshot();
    });

    it('should not render when open prop is falsy', () => {
        const { container } = dialogBoxRenderer({});
        expect(container).toBeEmptyDOMElement();
    });

    it('should call onCloseModal on pressing escape key', () => {
        const mockOnCloseModal = jest.fn();
        const { container } = dialogBoxRenderer({ open: true, onCloseModal: mockOnCloseModal });
        fireEvent.keyDown(container, { key: 'Escape', code: 27 });
        expect(mockOnCloseModal).toBeCalled();
    });

    it('should call onCloseModal on click on overlay background', () => {
        const mockOnCloseModal = jest.fn();
        const { container } = dialogBoxRenderer({ open: true, onCloseModal: mockOnCloseModal, shouldCloseOnOutsideClick: true });
        fireEvent.click(container.querySelector('#medly-dialog-box') as HTMLDivElement);
        expect(mockOnCloseModal).toBeCalled();
    });

    it('should be able to render any JSX element in header', () => {
        const mockOnCloseModal = jest.fn();
        const { container } = render(
            <DialogBox open onCloseModal={mockOnCloseModal}>
                <DialogBox.Header>Demo Header</DialogBox.Header>
                <DialogBox.Content>Demo Content</DialogBox.Content>
                <DialogBox.Actions>
                    <p>Demo Header</p>
                </DialogBox.Actions>
            </DialogBox>
        );
        expect(container.querySelector('p')).toBeInTheDocument();
    });
});
