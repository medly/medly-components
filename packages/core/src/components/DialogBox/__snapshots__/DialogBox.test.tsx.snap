// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialogBox component should render properly when it is open 1`] = `
.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  z-index: 10;
  margin-top: 3rem;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 > * {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c4 > * + * {
  margin: 1.6rem 0 0;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  box-sizing: border-box;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: rgba(67,84,101,0.65);
  z-index: 1000;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-animation: bgFadeIn 0.4s cubic-bezier(0,0,0.33,1);
  animation: bgFadeIn 0.4s cubic-bezier(0,0,0.33,1);
  overflow: hidden;
}

.c2 {
  background-color: #ffffff;
  color: #13181D;
  font-weight: 500;
  margin-bottom: 2.4rem;
}

.c1 {
  background: #ffffff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  box-shadow: 0 0.4rem 3.2 rgba(67,84,101,0.5);
  box-sizing: border-box;
  border-radius: 1.2rem;
  overflow-y: auto;
  max-height: 80%;
  width: calc(100% - 4.8rem);
  padding: 2.4rem;
  -webkit-animation: slideIn 0.4s cubic-bezier(0,0,0.33,1);
  animation: slideIn 0.4s cubic-bezier(0,0,0.33,1);
}

@media (min-width:600px) {
  .c4 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }

  .c4 > * {
    max-width: 50%;
  }

  .c4 > * + * {
    margin: 0 0 0 1.6rem;
  }
}

@media (min-width:608px) {
  .c1 {
    max-width: 56rem;
    padding: 3.2rem;
  }

  .c1::-webkit-scrollbar {
    width: 0.5rem;
  }

  .c1::-webkit-scrollbar-track {
    background-color: transparent;
  }

  .c1::-webkit-scrollbar-thumb {
    border-radius: 1rem;
    background-color: #b0bcc8;
  }
}

<div>
  <div
    class="c0"
    id="medly-dialog-box"
    open=""
  >
    <div
      class="c1"
      id="medly-dialog-box-popup"
      open=""
    >
      <div
        class="c2"
        id="medly-dialog-box-header"
      >
        <p>
          Demo Header
        </p>
      </div>
      <div
        class="c3"
        id="medly-dialog-box-content"
      >
        Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum
      </div>
      <div
        class="c4"
        id="medly-dialog-box-actions"
      >
        Demo Actions
      </div>
    </div>
  </div>
</div>
`;

exports[`DialogBox component should render properly with close icon when it is open 1`] = `
.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  z-index: 10;
  margin-top: 3rem;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c6 > * {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c6 > * + * {
  margin: 1.6rem 0 0;
}

.c5 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  box-sizing: border-box;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: rgba(67,84,101,0.65);
  z-index: 1000;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-animation: bgFadeIn 0.4s cubic-bezier(0,0,0.33,1);
  animation: bgFadeIn 0.4s cubic-bezier(0,0,0.33,1);
  overflow: hidden;
}

.c4 {
  background-color: #ffffff;
  color: #13181D;
  font-weight: 500;
  margin-bottom: 2.4rem;
}

.c1 {
  background: #ffffff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  box-shadow: 0 0.4rem 3.2 rgba(67,84,101,0.5);
  box-sizing: border-box;
  border-radius: 1.2rem;
  overflow-y: auto;
  max-height: 80%;
  width: calc(100% - 4.8rem);
  padding: 2.4rem;
  -webkit-animation: slideIn 0.4s cubic-bezier(0,0,0.33,1);
  animation: slideIn 0.4s cubic-bezier(0,0,0.33,1);
}

.c2 {
  overflow: visible;
  font-size: 2.4rem;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  cursor: pointer;
  padding: 0.8rem;
  border-radius: 50%;
  background-color: #EBF1FA;
}

.c2 * {
  fill-opacity: 1;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  fill: #607890;
}

.c3 {
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-perspective: 1000;
  -moz-perspective: 1000;
  -ms-perspective: 1000;
  perspective: 1000;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  top: 1.4rem;
  right: 1.4rem;
  z-index: 20;
  position: absolute;
  border-radius: 50%;
  border: none;
  background-color: #ffffff;
}

.c3 * {
  fill: #607890;
}

.c3:hover {
  background-color: #eff2f4;
  box-shadow: 0 2px 8px rgba(239,242,244,0.5);
}

.c3:hover * {
  fill: #607890;
}

.c3:active {
  background-color: #dfe4e9;
  box-shadow: none;
}

.c3:active * {
  fill: #435465;
}

@media (min-width:600px) {
  .c6 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }

  .c6 > * {
    max-width: 50%;
  }

  .c6 > * + * {
    margin: 0 0 0 1.6rem;
  }
}

@media (min-width:608px) {
  .c1 {
    max-width: 56rem;
    padding: 3.2rem;
  }

  .c1::-webkit-scrollbar {
    width: 0.5rem;
  }

  .c1::-webkit-scrollbar-track {
    background-color: transparent;
  }

  .c1::-webkit-scrollbar-thumb {
    border-radius: 1rem;
    background-color: #b0bcc8;
  }
}

<div>
  <div
    class="c0"
    id="medly-dialog-box"
    open=""
  >
    <div
      class="c1"
      id="medly-dialog-box-popup"
      open=""
    >
      <svg
        class="c2 c3"
        data-testid="medly-dialog-box-close-button"
        fill="none"
        height="1em"
        id="medly-dialog-box-close-button"
        title="medly-dialog-box-close-icon"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18.3 5.71a.996.996 0 0 0-1.41 0L12 10.59 7.11 5.7A.996.996 0 1 0 5.7 7.11L10.59 12 5.7 16.89a.996.996 0 1 0 1.41 1.41L12 13.41l4.89 4.89a.996.996 0 1 0 1.41-1.41L13.41 12l4.89-4.89c.38-.38.38-1.02 0-1.4z"
          fill="#000"
          fill-opacity=".54"
        />
      </svg>
      <div
        class="c4"
        id="medly-dialog-box-header"
      >
        <p>
          Demo Header
        </p>
      </div>
      <div
        class="c5"
        id="medly-dialog-box-content"
      >
        Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum
      </div>
      <div
        class="c6"
        id="medly-dialog-box-actions"
      >
        Demo Actions
      </div>
    </div>
  </div>
</div>
`;
